src/rdkafka_broker.c

```diff
@@ -1283,6 +1285,113 @@ rd_kafka_broker_random0 (const char *func, int line,
                                 rk, rd_false, state, NULL, filter, opaque)
 
 
+/**
+ * @returns the broker (with refcnt increased) with the highest weight based
+ *          based on the provided weighing function.
+ *
+ * If multiple brokers share the same weight reservoir sampling will be used
+ * to randomly select one.
+ *
+ * @param weight_cb Weighing function that should return the sort weight
+ *                  for the given broker.
+ *                  Higher weight is better.
+ *                  A weight of <= 0 will filter out the broker.
+ *                  The passed broker object is locked.
+ * @param features (optional) Required broker features.
+ *
+ * @locks_required rk(read)
+ * @locality any
+ */
+static rd_kafka_broker_t *
+rd_kafka_broker_weighted (rd_kafka_t *rk,
+                          int (*weight_cb) (rd_kafka_broker_t *rkb),
+                          int features) {
+        rd_kafka_broker_t *rkb, *good = NULL;
+        int highest = 0;
+        int cnt = 0;
+
+        TAILQ_FOREACH(rkb, &rk->rk_brokers, rkb_link) {
+                int weight;
+
+                rd_kafka_broker_lock(rkb);
+                if (features && (rkb->rkb_features & features) != features)
+                        weight = 0;
+                else
+                        weight = weight_cb(rkb);
+                rd_kafka_broker_unlock(rkb);
+
+                if (weight <= 0 || weight < highest)
+                        continue;
+
+                if (weight > highest) {
+                        highest = weight;
+                        cnt = 0;
+                }
+
+                /* If same weight (cnt > 0), use reservoir sampling */
+                if (cnt < 1 || rd_jitter(0, cnt) < 1) {
+                        if (good)
+                                rd_kafka_broker_destroy(good);
+                        rd_kafka_broker_keep(rkb);
+                        good = rkb;
+                }
+                cnt++;
+        }
+
+        return good;
+}
+
+/**
+ * @brief Weighing function to select a usable broker connections,
+ *        promoting connections according to the scoring below.
+ *
+ * Scoring:
+ *  2000 * is not a bootstrap broker
+ *  1000 + least idle last 10 minutes (unless blocking)
+ *  100 + least idle hours (if above 10 minutes idle)
+ *  10 * is not a logical broker
+ *  1 * is not blocking
+ *
+ * Will prefer the most recently used broker connection for two reasons:
+ * - this connection is most likely to function properly.
+ * - allows truly idle connections to be killed by the broker's/LB's
+ *   idle connection reaper.
+ *
+
+ * Connection must be up.
+ *
+ * @locks_required rkb
+ */
+static int rd_kafka_broker_weight_usable (rd_kafka_broker_t *rkb) {
+        int weight = 0;
+
+        if (!rd_kafka_broker_state_is_up(rkb->rkb_state))
+                return 0;
+
+        weight += 2000 * (rkb->rkb_nodeid != -1); /* is not a bootstrap */
+        weight += 10 * !RD_KAFKA_BROKER_IS_LOGICAL(rkb);
```

Comment:
should you return 0 for logical brokers since they are just an indirection on another broker connection, which will be considered by `rd_kafka_broker_weighted` anyway?

Reply:
Logical brokers are dedicated real connections, they're not an indirection to another broker_t, but rather an indirection to a variable hostname/address which is updated when e.g. the group coordinator is changing.
This is to say they are fully functional standard connections and the only reason not to give them as high of a weigh as standard connections is that the logical connections are dedicated for a certain purpose and should preferably be left unused for other means, but it is better to use an idle logical connection for a metadata query than to create a new connection on a disconnected broker thread.